src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/ValuePoolMutatorFactory.java

```diff
@@ -0,0 +1,186 @@
+  @Target({TYPE_USE})
+  @Retention(RUNTIME)
+  private @interface ValuePoolMarker {}
+
+  public static final Annotation VALUE_POOL_MARKER =
+      new TypeHolder<@ValuePoolMarker String>() {}.annotatedType()
+          .getAnnotation(ValuePoolMarker.class);
+
+  private final ValuePoolRegistry valuePoolRegistry;
+
+  ValuePoolMutatorFactory(ValuePoolRegistry valuePoolRegistry) {
+    this.valuePoolRegistry = valuePoolRegistry;
+  }
+
+  @Override
+  public Optional<SerializingMutator<?>> tryCreate(
+      AnnotatedType type, ExtendedMutatorFactory factory) {
+    if (type.getAnnotation(ValuePoolMarker.class) != null) {
+      return Optional.empty();
+    }
+    AnnotatedType markedType = TypeSupport.withExtraAnnotations(type, VALUE_POOL_MARKER);
+    return factory
+        .tryCreate(markedType)
+        .map(mutator -> ValuePoolMutator.wrapIfValuesExist(markedType, mutator, valuePoolRegistry));
+  }
+
+  private static final class ValuePoolMutator<T> extends SerializingMutator<T> {
+    private final SerializingMutator<T> mutator;
+    private final List<T> userValues;
+    private final double poolUsageProbability;
+
+    ValuePoolMutator(
+        SerializingMutator<T> mutator, List<T> userValues, double poolUsageProbability) {
+      this.mutator = mutator;
+      this.userValues = userValues;
+      this.poolUsageProbability = poolUsageProbability;
+    }
+
+    @SuppressWarnings("unchecked")
+    static <T> SerializingMutator<T> wrapIfValuesExist(
+        AnnotatedType type, SerializingMutator<T> mutator, ValuePoolRegistry valuePoolRegistry) {
+
+      if (valuePoolRegistry == null) {
+        return mutator;
+      }
+
+      Optional<Stream<?>> rawUserValues = valuePoolRegistry.extractRawValues(type);
+      if (!rawUserValues.isPresent()) {
+        return mutator;
+      }
+
+      List<T> userValues =
+          rawUserValues
+              .get()
+              .filter(value -> isSerializationStable(mutator, value))
+              .map(value -> mutator.detach((T) value))
+              .collect(Collectors.toList());
+
+      if (userValues.isEmpty()) {
+        return mutator;
+      }
+
+      double p = valuePoolRegistry.extractFirstProbability(type);
+      return new ValuePoolMutator<>(mutator, userValues, p);
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T> boolean isSerializationStable(SerializingMutator<T> mutator, Object value) {
+      ByteArrayOutputStream byteStream = new ByteArrayOutputStream();
+      try {
+        mutator.write((T) value, new DataOutputStream(byteStream));
+        byte[] originalSerialized = byteStream.toByteArray();
+        byteStream.reset();
+
+        T roundTrip =
+            mutator.read(new DataInputStream(new ByteArrayInputStream(originalSerialized)));
+        mutator.write(roundTrip, new DataOutputStream(byteStream));
+        byte[] roundTripSerialized = byteStream.toByteArray();
+        return Arrays.equals(originalSerialized, roundTripSerialized);
```

Comment:
For what mutator types do we need the roundtrip check? Wouldn't it be enough to check if the `mutator.write(...)` call does not throw, to filter out e.g. `CachedConstructorMutator`s?